### ISSUE TITLE
5-6x speed regression on large files (flake8 2.6.2 vs flake8 3.0.4)

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#166]
In GitLab by @asottile on Oct 3, 2016, 14:21

If you need more information please ask! I think this report is pretty thorough and I've been able to reproduce.  The fake file I create is not the actual file that is causing problems, but a representative file (the actual file is causing a nearly 20x slowdown)

## Version information

```
$ ./venv/bin/flake8  --version
2.6.2 (pycodestyle: 2.0.0, pyflakes: 1.2.3, mccabe: 0.5.2) CPython 2.7.6 on Linux

$ ./virtualenv_run/bin/flake8 --version
3.0.4 (mccabe: 0.5.2, pyflakes: 1.2.3, pycodestyle: 2.0.0) CPython 2.7.6 on Linux
```

## Make a terribly long file

```python
with open('f.py', 'w') as f:
    f.write('x = {\n')
    for _ in range(10000):
        f.write("    ('aaaaaaaaaaaaaaaaaaaaaaaaa', 'bbbbbbbbbbb'): 'ccccc',\n")
    f.write('}\n')
```

## Time difference

```
$ python make_test.py
$ time ./venv/bin/flake8 -j1 f.py

real	0m1.421s
user	0m1.336s
sys	0m0.076s
$ time ./virtualenv_run/bin/flake8 -j1 f.py

real	0m8.433s
user	0m8.304s
sys	0m0.116s
```

## Profiling

Generating a cprofile of this, I've noticed in the new version of flake8 it spends ~65% of the time in `flake8.checker:592:handle_newline` with most of that dominated by `pycodestyle:930:compound_statements` (47% total time) and `pycodestyle:404:continued_indentation` (10% total time)

I've attached a screenshot of this rendered w/ dot

![Screen_Shot_2016-10-03_at_2.20.28_PM](/uploads/722b974db24d59a767a53ca9888f8b0e/Screen_Shot_2016-10-03_at_2.20.28_PM.png)

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 25, 2016, 17:45

Hey @asottile ! 

Thanks for the detail in this! It's fantastic. I've been unwell for a while and am only just getting around to addressing Flake8 bugs created in my absence. It seems' like most of `handle_newline`'s time is spent running the logical line checks (`run_logical_checks`) and most of that time (~87%) is spent in pycodestyle. It would seem to me that we might want to optimize those, but also figure out why those weren't an issue prior to Flake8 taking control of running checks.

I'll take a look into this starting tomorrow probably.

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 21, 2018, 07:52

I'm not able to reproduce my original findings in python3 (similar in 2.7.15):

```console
$ flake8 --version
2.6.2 (pycodestyle: 2.0.0, mccabe: 0.5.3, pyflakes: 1.2.3) CPython 3.6.5 on Linux
$ time flake8 -j1 f.py 

real	0m7.612s
user	0m7.547s
sys	0m0.044s


$ flake8 --version
3.0.4 (mccabe: 0.5.3, pyflakes: 1.2.3, pycodestyle: 2.0.0) CPython 3.6.5 on Linux
$ time flake8 -j1 f.py 

real	0m8.545s
user	0m8.465s
sys	0m0.060s


$ flake8 --version
3.5.0 (mccabe: 0.6.1, pycodestyle: 2.3.1, pyflakes: 1.6.0) CPython 3.6.5 on Linux
$ time flake8 -j1 f.py 

real	0m5.263s
user	0m5.096s
sys	0m0.086s
```

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 21, 2018, 07:52

closed
